lib/krb5/get_default_principal.c

```diff
@@ -60,39 +60,38 @@ KRB5_LIB_FUNCTION krb5_error_code KRB5_LIB_CALL
 _krb5_get_default_principal_local (krb5_context context,
 				   krb5_principal *princ)
 {
-    krb5_error_code ret;
     const char *user;
-    uid_t uid;
+    const char *second_component = NULL;
 
     *princ = NULL;
 
-    uid = getuid();
-    if(uid == 0) {
+    if (!issuid() && getuid() == 0) {
+        /* We're root or root shell spawned via sudo or similar */
+        /* XXX Use getlogin_r() instead; make an rk_getlogin_r() */
 	user = getlogin();
-	if(user == NULL)
-	    user = get_env_user();
 	if(user != NULL && strcmp(user, "root") != 0)
-	    ret = krb5_make_principal(context, princ, NULL, user, "root", NULL);
+            /* <username>/root@<DEFAULT_REALM> */
+	    second_component = "root";
 	else
-	    ret = krb5_make_principal(context, princ, NULL, "root", NULL);
+            /* root@<DEFAULT_REALM> */
+            user = "root";
+    } else if (issuid()) {
+        /* We're a set-uid process -- don't trust the environ, only utmp */
+	user = getlogin(); /* XXX use roke_get_username() */
     } else {
-	struct passwd *pw = getpwuid(uid);
-	if(pw != NULL)
-	    user = pw->pw_name;
-	else {
-	    user = get_env_user();
-	    if(user == NULL)
-		user = getlogin();
-	}
-	if(user == NULL) {
-	    krb5_set_error_message(context, ENOTTY,
-				   N_("unable to figure out current "
-				      "principal", ""));
-	    return ENOTTY; /* XXX */
-	}
-	ret = krb5_make_principal(context, princ, NULL, user, NULL);
+        user = get_env_user();
+        if (user == NULL)
+            user = getlogin();
+    }
+
+    if (user == NULL) {
```

Comment:
We should probably still try getpwuid() if get_env_user() and getlogin() borth return NULL. It used to be tried first, and should not disappear entirely as ptys don't necessary have accurate wtmp entries.